### ISSUE TITLE
[EPIC] Support additional parameters input for JDBC sources & destinations

### ISSUE THREAD

**Issue by tuliren** [airbytehq/airbyte#9501]
## Tell us about the problem you're trying to solve
- JDBC support appending additional parameters in the URL. This enables users to customize their connection.
- Some of our JDBC connectors support this feature, but some do not.

## Describe the solution you’d like
- Support this feature in all JDBC based connectors

## Describe the alternative you’ve considered or used
N/A

## TODOs

- [ ] Support additional parameters for high priority connectors.
- [ ] Support additional parameters for medium priority connectors.

| Connector | Source | Priority | Destination | Priority | Note |
| --- | :---: | :---: | :---: | :---: | --- |
| MySQL | ✅ | - | ✅ #9434 | High | |
| SQL Server | ❌ | High | ❌ | High | |
| Postgres | ❌ | High | ❌ | High | |
| Redshift | ❌ | Medium | ❌ | High | |
| Snowflake | ✅ #9623 | Medium | ✅ #9623 | High | |
| Oracle | ❌ | Medium | ❌ | Medium | |
| DB2 | ❌ | Low | No such connector | - | |
| Clickhouse | ❌ | Low | ❌ | Low | |
| CockroachDB | ❌ | Low | No such connector | - | |

## Relevant Issues
- https://github.com/airbytehq/airbyte/issues/8785
- https://github.com/airbytehq/airbyte/issues/9434
- https://github.com/airbytehq/airbyte/issues/9467
  - follow-up with https://github.com/airbytehq/airbyte/issues/9953
- https://github.com/airbytehq/airbyte/issues/10593
- https://github.com/airbytehq/airbyte/issues/10717
- https://github.com/airbytehq/airbyte/issues/10718
- https://github.com/airbytehq/airbyte/issues/10719
- https://github.com/airbytehq/airbyte/issues/10720
- https://github.com/airbytehq/airbyte/issues/10721
- https://github.com/airbytehq/airbyte/issues/10722
- https://github.com/airbytehq/airbyte/issues/10723
- https://github.com/airbytehq/airbyte/issues/10724
- https://github.com/airbytehq/airbyte/issues/10725
- https://github.com/airbytehq/airbyte/issues/10726
- https://github.com/airbytehq/airbyte/issues/10727
- https://github.com/airbytehq/airbyte/issues/10728
- https://github.com/airbytehq/airbyte/issues/10729
- https://github.com/airbytehq/airbyte/issues/10730

**Comment by noahkawasakigoogle** [comment]
🎉

**Comment by noahkawasakigoogle** [comment]
Another benefit to this I'm realizing, for a number of JDBC based connectors, there are "StrictEncrypted" versions of those connectors which basically re-use the entire connector but manipulate the SSL JDBC parameter. By allowing any params to be added in, theoretically the base connector could have any SSL behavior desired and some duplicate connectors could be deleted.

**Comment by noahkawasakigoogle** [comment]
Hey @sherifnada I can reply to these with my thoughts because I've had to think about the same scenarios:

1. It would be a lot of overhead for every connector to have targeted options for every possible parameter, as there would need to be potentially 20+ options for each JDBC based connector. Every time someone realizes they want another option exposed to be changeable, another PR would be needed, and another release, and another update. As drivers add new features, Airbyte must keep up with constant updates. It wouldn't be hard code-wise. But it's a lot of work for Airbyte to commit to over time, especially with an infinite amount of connectors some day. Allowing user set parameters solves this once and for all. 
2. Documenting certain required or "suggested" parameter values and known issues could solve this, but its definitely a real problem. Things like SSL, schema/database settings, are very common and mostly uniform across JDBC drivers, having Airbyte expose those as generic options feels right. While uncommon driver specific things like Snowflake's QUERY_TAG could be flexibly set with this. 
3. This could be as simple as wrapping each parameter Airbyte wants to set a value for with:

```
// If not already set, set this default
if (!config.get("jdbc_url_params").contains("some_param")) {
  jdbcUrl.append("&some_param=aribyte_default_value")
}
```

And then adding a test suite that asserts it's possible to override things.

I personally did _not_ include that in the Snowflake PR I put up because those two parameters were things that, in my opinion and reading of the comments, should always have those values (but this should be documented now that I think about it).

A con about this approach I'll point out is the security vulnerabilities with allowing arbitrary parameters be set in this string. As JDBC URL's are a common target for attacks by specifying certain things like `allowLocalInfile` for MYSQL allowing reading host filesystem files. However this issue exists currently with the host/database or any other string field so Airbyte will eventually need to add some restrictions someday anyways.

**Comment by tuliren** [comment]
Noah, thank you for the thorough reply.

The only thing to add is that if there are some parameters that should not be set by the user for security concerns, we can document it, ignore the user input just for those parameters, and writing a warning in the log.

**Comment by alafanechere** [comment]
> The only thing to add is that if there are some parameters that should not be set by the user for security concerns, we can document it, ignore the user input just for those parameters, and writing a warning in the log.

@tuliren do you think this is a blocker for the merge of Noah's PR's https://github.com/airbytehq/airbyte/pull/9623? For Snowflake, I can't see any hardcoded JDBC params that could be overwritten by the user's input.

**Comment by sherifnada** [comment]
Assigning to @tuliren for scoping a generalized solution

**Comment by sherifnada** [comment]
@luisazuluaga we'll share an ETA once we know it!

**Comment by tuliren** [comment]
This can be the second project for Alexandre. Here is the project [doc](https://docs.google.com/document/d/12qy6OHcGmsv7FzPLLNK_iixfsriz7_UktoAK3jUDaoA/edit#) that adds support for additional parameters for all JDBC destinations.

**Comment by girarda** [comment]
I'm working on this right now.
ETA for a generic solution is by end of week. PR is https://github.com/airbytehq/airbyte/pull/10421
Once that's merged, we'll only need to update the connectors spec

**Comment by lfzuluagah** [comment]
@girarda @tuliren  one of my team members finished the MSSQL source with additional URL params, let us know if it's appropriate to generate a PR request to review this contribution.

**Comment by girarda** [comment]
@lfzuluagah a PR would be great! Feel free to submit one at your convenience

**Comment by philippeboyd** [comment]
@lfzuluagah any news on the PR from your team mate?

**Comment by lfzuluagah** [comment]
Daniel, any pending on this PR?
________________________________
De: Philippe Boyd ***@***.***>
Enviado: lunes, 25 de julio de 2022 10:07 a. m.
Para: airbytehq/airbyte ***@***.***>
Cc: Luisa Zuluaga ***@***.***>; Mention ***@***.***>
Asunto: Re: [airbytehq/airbyte] [EPIC] Support additional parameters input for JDBC sources & destinations (Issue #9501)


@lfzuluagah<https://nam11.safelinks.protection.outlook.com/?url=https%3A%2F%2Fgithub.com%2Flfzuluagah&data=05%7C01%7C%7C43b8ca56ae384edda95c08da6e4f5cbf%7C6727db764b684a92956f1b9db0c13391%7C0%7C0%7C637943584391149787%7CUnknown%7CTWFpbGZsb3d8eyJWIjoiMC4wLjAwMDAiLCJQIjoiV2luMzIiLCJBTiI6Ik1haWwiLCJXVCI6Mn0%3D%7C3000%7C%7C%7C&sdata=NnW1FP4%2BvKW3HSoVtVTKruEwRh4z4kr717c8lu2TCII%3D&reserved=0> any news on the PR from your team mate?

—
Reply to this email directly, view it on GitHub<https://nam11.safelinks.protection.outlook.com/?url=https%3A%2F%2Fgithub.com%2Fairbytehq%2Fairbyte%2Fissues%2F9501%23issuecomment-1194174262&data=05%7C01%7C%7C43b8ca56ae384edda95c08da6e4f5cbf%7C6727db764b684a92956f1b9db0c13391%7C0%7C0%7C637943584391149787%7CUnknown%7CTWFpbGZsb3d8eyJWIjoiMC4wLjAwMDAiLCJQIjoiV2luMzIiLCJBTiI6Ik1haWwiLCJXVCI6Mn0%3D%7C3000%7C%7C%7C&sdata=xKaGnmmC7B4Id%2Bx9fPGoh7CAMEhjqsAX0IIcvhiyKOQ%3D&reserved=0>, or unsubscribe<https://nam11.safelinks.protection.outlook.com/?url=https%3A%2F%2Fgithub.com%2Fnotifications%2Funsubscribe-auth%2FAMQHIXX6UCGFW6EGKHVRRU3VV2USHANCNFSM5L6BIBCQ&data=05%7C01%7C%7C43b8ca56ae384edda95c08da6e4f5cbf%7C6727db764b684a92956f1b9db0c13391%7C0%7C0%7C637943584391149787%7CUnknown%7CTWFpbGZsb3d8eyJWIjoiMC4wLjAwMDAiLCJQIjoiV2luMzIiLCJBTiI6Ik1haWwiLCJXVCI6Mn0%3D%7C3000%7C%7C%7C&sdata=Z51bnKdKpOKOt9ARXI7z6SjWhTqMarLad5RnhRVW8rc%3D&reserved=0>.
You are receiving this because you were mentioned.Message ID: ***@***.***>

________________________________
Disclaimer: This message and information contained in or attached to this communication is privileged and confidential and intended only for the person named above. If you are not the intended recipient of this transmission, you are hereby notified that any dissemination, distribution or copying of this communication to anyone other than the intended recipient is strictly prohibited. If you received this communication in error, please do not read it. Please immediately reply to the sender that you have received this communication in error and delete this communication from your computer. Thank you.
________________________________

**Comment by philippeboyd** [comment]
I've encountered a couple of commits in the code potentially related to this feature since mid-July: 5edf3fed7189fe13531468fad64998787e50ef49 and 063b01f51c0bf3b5702156c8aee407ec5396016d; both made by @ryankfu

Wondering if this ever made it to the connectors?

When I try to set the JDBC param `queryTimeout=0` to disable query timeout, I still get a MSSQL query timeout after 10 minutes which seems to be the default server side. See issue and comment on https://github.com/airbytehq/airbyte/issues/14459#issuecomment-1194175071

**Comment by evantahler** [comment]
Closing! 
All the linked issues appear to be complete
